scanner/decisions/0010-separate-versioning-for-vuln-bundles.md

```diff
@@ -0,0 +1,82 @@
+
+## Context
+
+A Vulnerability Bundle is an archive containing updates to vulnerability data in Scanner.  They are constantly updated by the Vulnerability Exporter (aka. Exporter), which runs periodically in GitHub's Workflows and uses ClairCore to fetch the security data to convert it to a JSON-based schema.  Then, the workflow bundles the data into a ZIP archive.  The schema is owned by ClairCore.
+
+A Scanner instance runs the Vulnerability Updater (aka. Updater) to fetch the bundle and use ClairCore to parse the schema and update the vulnerability database.  Suppose ClairCore changes the schema in a backward-incompatible manner, and the Exporter updates the bundle.  In that case, existing Scanner instances cannot consume the data.  A Scanner version bump to a version with a ClairCore version supporting the new schema would be required.  Until that bump happens, no vulnerability update will occur. 
+
+To address that scenario, Scanner created one bundle per Stackrox release.  Each bundle is created with an Exporter running the same ClairCore version used to build the release, ensuring compatibility.  Bundles are served at `https://definitions.stackrox.io/v4/V/vulnerabilities.zip` where `V` will be a Stackrox release version such as "4.5.0".  Scanner uses its release version tagged during build time to determine which version to fetch from.  The versions to build bundles for are taken from a file named `scanner/updater/version/RELEASE_VERSION`.
+
+For its purpose, this approach needs to be revised.
+
+1. The vulnerability schema rarely changes, but maintaining multiple bundles is expensive.  It requires multiple GitHub workflow runs, which increases the chances of failure and pressure on the security sources' endpoints.
+2. It isn't easy to backport fixes and updates to previous releases.  This requires patch releases, which trigger the release workflow.  Improving the release process is noisy, convoluted, and complex.
+3. The offline mode's "latest bundle" needs to exist as a complex solution to help users transition to the versioned bundle.  It contains all versions in a single bundle.  But one bundle per release makes the latest bundle large and, most of the time, essentially carries the same data since the schema is the same.
+
+## Decision
+
+We will implement a versioning stream for vulnerability bundles, bumped when we determine that ClairCore's schema has changed.
+
+Initially, two version streams will exist (table 1 below):
+
+| Stream | Git Ref      | Description                                                        |
+|--------|--------------|--------------------------------------------------------------------|
+| dev    | heads/master | This stream tracks the latest vulnerability schema                 |
+| v1     | heads/4.4.5  | This stream tracks the schema used by the current Stackrox release |
+
+On the subsequent releases (for example,. `4.4.6`), one of the two will be carried out:
```

Comment:
Could you please clarify if this step needs to be manually performed by the Scanner team, or will it be included as part of the release process? 
Could the workflow be briefly discussed in this ADR?

For example, if we have a new formal release (4.6.0) with a schema change and a patch release (4.4.6) is issued, how would this workflow operate?

Reply:
> Could you please clarify if this step needs to be manually performed by the Scanner team, or will it be included as part of the release process?

Yes, it should be included in the release process.

> For example, if we have a new formal release (4.6.0) with a schema change and a patch release (4.4.6) is issued, how would this workflow operate?

Mapping file before 4.6.0:

| Bundle Version | Git Reference |
| - | - |
| dev | heads/master |
| v1 | heads/4.4.5 |

Mapping file after 4.6.0:

| Bundle Version | Git Reference |
| - | - |
| dev | heads/master |
| v1 | heads/4.4.5 |
| v2 | heads/4.6.0 |

Mapping file after the patch release (notice: it's the same as above):

| Bundle Version | Git Reference |
| - | - |
| dev | heads/master |
| v1 | heads/4.4.5 |
| v2 | heads/4.6.0 |

Assuming the patch release `4.4.6` didn't adopt the new schema changes. If that's the case, then we would need to change the `VULNERABILITY_BUNDLE_VERSION` file of `4.4.6` to "v2". Notice we would still build v1 to make sure people using `4.4.5` would still get updates.

Does that clarify things for you?